### PR TITLE
chore: account offline

### PR DESCRIPTION
**Pull request by xiaolang124**
1. Do not clear account cache when offline.
2. TreeView will show m365 account name when user has logged in.

### CODE REVIEW

File: packages/vscode-extension/src/commonlib/codeFlowLogin.ts [hunk 1]
```diff
@@ -356,3 +360,16 @@ export function ConvertTokenToJson(token: string): object {
   const buff = Buffer.from(array[1], "base64");
   return JSON.parse(buff.toString(UTF8));
 }
+
+export async function checkIsOnline(): Promise<boolean> {
+  try {
+    await new Promise<http.IncomingMessage>((resolve, reject) => {
+      const url = "https://login.microsoftonline.com/";
+      https.get(url, resolve).on("error", reject);
```
Comment (swatDong) [inline packages/vscode-extension/src/commonlib/codeFlowLogin.ts:368]: Does `get` automatically follow redirect? A full `get` could be heavy, any `head` option?

File: packages/vscode-extension/src/commonlib/codeFlowLogin.ts [hunk 2]
```diff
@@ -258,6 +259,9 @@ export class CodeFlowLogin {
                   error.message
                 )
             );
+            if (!(await checkIsOnline())) {
```
Comment (swatDong) [inline packages/vscode-extension/src/commonlib/codeFlowLogin.ts:262]: Any change needed in CLI?

File: packages/vscode-extension/src/accountTree.ts [hunk 3]
```diff
@@ -312,6 +312,16 @@ export async function registerAccountTreeHandler(): Promise<Result<Void, FxError
...
+      } else if (accountInfo !== undefined) {
+        const treeItem = {
+          commandId: "fx-extension.signinM365",
+          label: (accountInfo.upn as string) ? (accountInfo.upn as string) : "",
+          callback: signinM365Callback,
+          parent: TreeCategory.Account,
+          contextValue: "signedinM365",
+          icon: "M365",
+        };
+        tools.treeProvider?.add([treeItem]);
```
Comment (HuihuiWu-Microsoft) [inline packages/vscode-extension/src/accountTree.ts:324]: I think here should be 'refresh()'?

Comment (HuihuiWu-Microsoft) [reply packages/vscode-extension/src/accountTree.ts:324]: BTW, what about showing a warning to remind user he is offline, or user might get confused the account shows normally but error occurs

Comment (xiaolang124) [reply packages/vscode-extension/src/accountTree.ts:324]: I have tried with ```refresh```, the account info will not show.

image

image

File: packages/vscode-extension/src/commonlib/codeFlowLogin.ts (hunk 2 above)
Comment (xiaolang124) [reply packages/vscode-extension/src/commonlib/codeFlowLogin.ts:262]: Already add CLI logic.

File: packages/vscode-extension/src/commonlib/codeFlowLogin.ts (hunk 1 above)
Comment (xiaolang124) [reply packages/vscode-extension/src/commonlib/codeFlowLogin.ts:368]: Changed to ```head``` request.

File: packages/vscode-extension/src/accountTree.ts (hunk 3 above)
Comment (HuihuiWu-Microsoft) [reply packages/vscode-extension/src/accountTree.ts:324]: It's weird, from the code logic, add() should be the one that doesn't work.

Comment (HuihuiWu-Microsoft) [reply packages/vscode-extension/src/accountTree.ts:324]: I got it, the item is added the first time extension started. The tree view also listened to SharepointLoginProvider, will spfx deploy affect the status?

Comment (xiaolang124) [reply packages/vscode-extension/src/accountTree.ts:324]: I think so, spfx login can also trigger treeView update user info.

Comment (HuihuiWu-Microsoft) [reply packages/vscode-extension/src/accountTree.ts:324]: Then is it needed to also update the sharepointLoginProvider getStatus() logic ?

Comment (xiaolang124) [reply packages/vscode-extension/src/accountTree.ts:324]: I update sharepoint logic, but it does not work. The treeProvider has not initialized when offline at this time.

image

Comment (xiaolang124) [reply packages/vscode-extension/src/accountTree.ts]: Move setStatus location to make sure notify will update treeView.

**Review by HuihuiWu-Microsoft — APPROVED** [review]

**Review by swatDong — APPROVED** [review]

**Review by Alive-Fish — APPROVED** [review]

**Pull request merged**